jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -212,52 +202,61 @@ public boolean tryExecute(Runnable task)
     {
         if (LOG.isDebugEnabled())
             LOG.debug("{} tryExecute {}", this, task);
-
         if (task == null)
             return false;
 
-        ReservedThread thread = _stack.pollFirst();
-        if (thread == null)
+        boolean offered = _queue.offer(task);
+        int size;
+        while (true)
         {
-            if (task != STOP)
-                startReservedThread();
-            return false;
+            long count = _count.get();
+            size = getLo(count);
+            if (size < 0 || !offered)
+                break;
+            if (_count.compareAndSet(count, getHi(count), --size))
+                break;
         }
-
-        int size = _size.decrementAndGet();
-        if (!thread.offer(task))
-            return false;
-
         if (size == 0 && task != STOP)
             startReservedThread();
-
-        return true;
+        return offered;
     }
 
     private void startReservedThread()
     {
-        try
+        while (true)
         {
-            while (true)
+            long count = _count.get();
+            int pending = getHi(count);
+            int size = getLo(count);
+            if (size < 0 || pending + size >= _capacity)
+                return;
+            if (size == 0)
+                _lastZero = System.nanoTime();
+            if (!_count.compareAndSet(count, pending + 1, size))
+                continue;
+
+            if (LOG.isDebugEnabled())
+                LOG.debug("{} startReservedThread p={}", this, pending + 1);
+            try
             {
-                // Not atomic, but there is a re-check in ReservedThread.run().
-                int pending = _pending.get();
-                int size = _size.get();
-                if (pending + size >= _capacity)
-                    return;
-                if (_pending.compareAndSet(pending, pending + 1))
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("{} startReservedThread p={}", this, pending + 1);
-                    _executor.execute(new ReservedThread());
-                    return;
-                }
+                ReservedThread thread = new ReservedThread();
+                _executor.execute(thread);
+                _threads.add(thread);
```

Comment:
I think these 2 lines should be swapped, as we want to record the thread before we submit it for execution.